extended_lengyel/extended_lengyel_model/forward_model.py

```diff
@@ -0,0 +1,358 @@
+        seed_c_z = impurity_fraction * CzLINT_for_seed_impurities.weights
+        fixed_c_z = CzLINT_for_fixed_impurities.weights
+        z_effective = (
+            starting_z_effective
+            + (seed_mean_z * (seed_mean_z - 1.0) * seed_c_z).sum(dim="dim_species")
+            + (fixed_mean_z * (fixed_mean_z - 1.0) * fixed_c_z).sum(dim="dim_species")
+        )
+        return z_effective.values
+
+    convergence: dict[str, Any] = dict(equal_nan=False, atol=0.0, rtol=1e-6)
+
+    def relax(new_value, prev_value, relaxation_factor=0.4):
+        return relaxation_factor * new_value + (1 - relaxation_factor) * prev_value
+
+    mu_0 = 1.2566370621250601e-6 # meter * tesla / ampere
+    elementary_charge = 1.602176634e-19 # coulomb
+    MA_to_A = 1.0e6 # MA / A
+    amu_to_kg = 1.6605390666e-27 # amu / kilogram
+    MW_to_W = 1.0e6 # MW / W
+    eV_to_J = elementary_charge
+    boltzmann_constant = 1.380649e-23 # joule/kelvin
+    n20_to_m3 = 1.0e20 # 10^20 / m^3 to 1 / m^3
+
+    # Convert all inputs to SI units, except for electron-volts
+    plasma_current = plasma_current * MA_to_A
+    target_angle_of_incidence = np.deg2rad(target_angle_of_incidence)
+    power_crossing_separatrix = power_crossing_separatrix * MW_to_W
+    separatrix_electron_density = separatrix_electron_density / n20_to_m3
+
+    shaping_factor = np.sqrt((1.0 + elongation_psi95**2 * (1.0 + 2.0 * triangularity_psi95**2 - 1.2 * triangularity_psi95**3)) / 2.0)
+    poloidal_circumference = 2.0 * np.pi * minor_radius * shaping_factor
+
+    upstream_toroidal_field = magnetic_field_on_axis * (major_radius / (major_radius + minor_radius))
+    separatrix_average_poloidal_field = mu_0 * plasma_current / poloidal_circumference
+    upstream_poloidal_field = ratio_of_upstream_to_average_poloidal_field * separatrix_average_poloidal_field
+
+    cylindrical_safety_factor = magnetic_field_on_axis / separatrix_average_poloidal_field * minor_radius / major_radius * shaping_factor
+
+    fieldline_pitch_at_omp = np.sqrt(upstream_toroidal_field**2 + upstream_poloidal_field**2) / upstream_poloidal_field
+
+    kappa_e0 = 2390.0 # W / (m * eV**3.5)
+    fraction_of_power_entering_flux_tube = (1.0 - 1.0 / np.e) * fraction_of_P_SOL_to_divertor
+
+    # Starting values for the iterative solver
+    target_electron_temp = 2.0 # eV
+    electron_temp_at_cc_interface = 2.5 # eV
+    divertor_entrance_electron_temp = 50.0 # eV
+    separatrix_electron_temp = 100.0 # eV
+    alpha_t = 0.0
+
+    prev_target_electron_temp = np.nan
+    prev_parallel_heat_flux_at_cc_interface = np.nan
+    prev_divertor_entrance_electron_temp = np.nan
+    prev_separatrix_electron_temp = np.nan
+    prev_alpha_t = np.nan
+
+    target_electron_temp_its = np.zeros(iterations)
+    parallel_heat_flux_at_cc_interface_its = np.zeros(iterations)
+    separatrix_electron_temp_its = np.zeros(iterations)
+    alpha_t_its = np.zeros(iterations)
+
+    for _it in range(iterations):
```

Comment:
I don't see a reason for the underscore prefix, but I may be missing something. But this is really a nit, it's fine either way.